azurelinuxagent/common/utils/shellutil.py

```diff
@@ -117,30 +117,54 @@ def _encode_command_output(output):
     return ustr(output, encoding='utf-8', errors="backslashreplace")
 
 
-def run_command(command):
+class CommandError(Exception):
     """
-    Wrapper for subprocess.Popen. Executes the given command and returns its stdout.
-    Logs any errors executing the command and raises an exception.
+    Exception raised by run_command when the command returns an error
     """
-    retcode = 0
+    @staticmethod
+    def _get_message(command, returncode):
+        command_name = command[0] if isinstance(command, list) and len(command) > 0 else command
+        return "'{0}' failed: {1}".format(command_name, returncode)
+
+    def __init__(self, command, returncode, stdout, stderr):
+        super(Exception, self).__init__(CommandError._get_message(command, returncode))
+        self.command = command
+        self.returncode = returncode
+        self.stdout = stdout
+        self.stderr = stderr
+
+
+def run_command(command, log_error=False):
+    """
+    Executes the given command and returns its stdout as a string.
+    If there are any errors executing the command it logs details about the failure and raises a RunCommandException;
+    if 'log_error' is True, it also logs details about the error.
+    """
+    def format_command(cmd):
+        return " ".join(cmd) if isinstance(cmd, list) else command
```

Comment:
`if isinstance(cmd, list) else command` 
you dont need this check everywhere, the join command would work even if its just a string
```python
a = ['a', 'b']
'.'.join(a)
'a.b'
a = 'a'
'.'.join(a)
'a'
' '.join(a)
'a'
```

Else everything else looks good

Reply:
> " ".join("/var/lib/waagent")
'/ v a r / l i b / w a a g e n t'

:)